server/src/main/java/org/elasticsearch/index/shard/GlobalCheckpointListeners.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.shard;
+
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.Executor;
+import java.util.function.LongSupplier;
+
+import static org.elasticsearch.index.seqno.SequenceNumbers.NO_OPS_PERFORMED;
+import static org.elasticsearch.index.seqno.SequenceNumbers.UNASSIGNED_SEQ_NO;
+
+/**
+ * Represents a collection of global checkpoint listeners. This collection can be added to, and all listeners present at the time of an
+ * update will be notified together. All listeners will be notified when the shard is closed.
+ */
+public class GlobalCheckpointListeners implements Closeable {
+
+    /**
+     * A global checkpoint listener consisting of a callback that is notified when the global checkpoint is updated or the shard is closed.
+     */
+    @FunctionalInterface
+    public interface GlobalCheckpointListener {
+        /**
+         * Callback when the global checkpoint is updated or the shard is closed. If the shard is closed, the value of the global checkpoint
+         * will be set to {@link org.elasticsearch.index.seqno.SequenceNumbers#UNASSIGNED_SEQ_NO} and the exception will be non-null. If the
+         * global checkpoint is updated, the exception will be null.
```

Comment:
Nit: I wonder if we should have an onFailure method here for all kind of failures and send the IndexShardClosedException down that route.  The down side is of course that people wouldn't be able to pass a method references, but the method won't need to start with if (e != null) etc.

Reply:
I prefer to use a functional interface for enabling the use of lambda expressions.